src/libraries/System.Security.Cryptography.Cose/tests/CoseSign1MessageTests.Sign.cs

```diff
@@ -0,0 +1,418 @@
+            {
+                ECDsa ecDsaKey = ECDsaKeysWithoutPrivateKey[(ECDsaAlgorithm)algorithm];
+                Assert.ThrowsAny<CryptographicException>(() => CoseSign1Message.Sign(content, ecDsaKey, new HashAlgorithmName(hashAlgorithm)));
+                key = ecDsaKey;
+            }
+            else
+            {
+                Assert.ThrowsAny<CryptographicException>(() => CoseSign1Message.Sign(content, RSAKeyWithoutPrivateKey, new HashAlgorithmName(hashAlgorithm)));
+                key = RSAKeyWithoutPrivateKey;
+            }
+
+            Assert.ThrowsAny<CryptographicException>(() => CoseSign1Message.Sign(content, GetHeaderMapWithAlgorithm(algorithm), GetEmptyHeaderMap(), key));
+        }
+
+        // SignWithUnsupportedKey (ECDA?)
+        [Fact]
+        public void SignWithUnsupportedKey()
+        {
+            AsymmetricAlgorithm key = DSA.Create();
+            // CoseHeader will say that a supported combination of key-algorithm will be used.
+            Assert.Throws<CryptographicException>(() => CoseSign1Message.Sign(GetDummyContent(ContentTestCase.Small), GetHeaderMapWithAlgorithm((int)ECDsaAlgorithm.ES256), GetEmptyHeaderMap(), key));
+        }
+
+        // Hash algorithms
+        [Theory]
+        [InlineData("SHA1")]
+        [InlineData("FOO")]
+        public void SignWithUnsupportedHashAlgorithm(string hashAlgorithm)
+        {
+            byte[] content = GetDummyContent(ContentTestCase.Small);
+
+            Assert.Throws<CryptographicException>(() => CoseSign1Message.Sign(GetDummyContent(ContentTestCase.Small), ECDsaKeys[ECDsaAlgorithm.ES256], new HashAlgorithmName(hashAlgorithm)));
+            Assert.Throws<CryptographicException>(() => CoseSign1Message.Sign(GetDummyContent(ContentTestCase.Small), RSAKey, new HashAlgorithmName(hashAlgorithm)));
+            Assert.Throws<CryptographicException>(() => CoseSign1Message.Sign(GetDummyContent(ContentTestCase.Small), GetHeaderMapWithAlgorithm(-47 /*ES256K*/), GetEmptyHeaderMap(), ECDsaKeys[ECDsaAlgorithm.ES256]));
+        }
+
+        [Theory]
+        [InlineData((int)ECDsaAlgorithm.ES256, CoseTestHelpers.SHA256)]
+        [InlineData((int)ECDsaAlgorithm.ES384, CoseTestHelpers.SHA384)]
+        [InlineData((int)ECDsaAlgorithm.ES512, CoseTestHelpers.SHA512)]
+        [InlineData((int)RSAAlgorithm.PS256, CoseTestHelpers.SHA256)]
+        [InlineData((int)RSAAlgorithm.PS384, CoseTestHelpers.SHA384)]
+        [InlineData((int)RSAAlgorithm.PS512, CoseTestHelpers.SHA512)]
+        public void SignWithSupportedHashAlgorithm(int algorithm, string hashAlgorithm)
+        {
+            byte[] content = GetDummyContent(ContentTestCase.Small);
+            AsymmetricAlgorithm key;
+
+            if (Enum.IsDefined(typeof(ECDsaAlgorithm), algorithm))
+            {
+                ECDsa ecDsaKey = ECDsaKeys[(ECDsaAlgorithm)algorithm];
+                Assert.NotNull(CoseSign1Message.Sign(content, ecDsaKey, new HashAlgorithmName(hashAlgorithm)));
```

Comment:
On the one hand, this is... a unit test.  It does one thing, measures one result, and exits.

On the other, if it fails then a lot is going to fail, and there's no test-sanity-dependency graph, so it has low value.

A more valueable version of this test would be

* Call Sign, save the result
* Assert.NotNull, that's fine.
* Verify everything about the results
  * Fire up a CBOR reader.
    * Check that we wrote the tag.
    * Check it's an array-4
    * Content looks correct (null/bstr-with-correct-value)
    * Check that we wrote the expected alg protected header, and nothing else.
    * Check that we wrote no unprotected headers
    * Check that the signature length makes sense
      * (rsa.KeySize + 7) / 8
      * (ecdsa.KeySize + 7) / 4 (because it's `(ecdsa.KeySize + 7) / 8 * 2`)

Reply:
Another way to check all of that is to take a message that already looks like that and check that all the bytes before the signature (but including the bstr identifier and payload length) match that existing value.

Reply:
This should be covered by `CoseTestHelpers.AssertSign1Message`. Thanks.